core/drivers/plic.c

```diff
@@ -0,0 +1,287 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright 2022 NXP
+ *
+ */
+
+#include <assert.h>
+#include <config.h>
+#include <io.h>
+#include <kernel/dt.h>
+#include <kernel/interrupt.h>
+#include <kernel/panic.h>
+#include <mm/core_memprot.h>
+#include <mm/core_mmu.h>
+#include <platform_config.h>
+#include <drivers/plic.h>
+#include <trace.h>
+
+#ifdef CFG_RISCV_WITH_PLIC
+
+#define PLIC_PRIORITY_OFFSET		0
+#define PLIC_PENDING_OFFSET		0x1000
+#define PLIC_ENABLE_OFFSET		0x2000
+#define PLIC_THRESHOLD_OFFSET		0x200000
+#define PLIC_CLAIM_OFFSET		0x200004
+
+#define PLIC_PRIORITY_SHIFT_PER_SOURCE	U(2)
+#define PLIC_PENDING_SHIFT_PER_SOURCE	U(0)
+
+#define PLIC_ENABLE_SHIFT_PER_TARGET	U(7)
+#define PLIC_THRESHOLD_SHIFT_PER_TARGET	U(12)
+#define PLIC_CLAIM_SHIFT_PER_TARGET	U(12)
+
+#define PLIC_PRIORITY(base, source) \
+	((base) + PLIC_PRIORITY_OFFSET +\
+	SHIFT_U32(source, PLIC_PRIORITY_SHIFT_PER_SOURCE))
+#define PLIC_PENDING(base, source) \
+	((base) + PLIC_PENDING_OFFSET +\
+	(4 * ((source) / 32)))
+#define PLIC_ENABLE(base, source, hart) \
+	((base) + PLIC_ENABLE_OFFSET +\
+	SHIFT_U32(hart, PLIC_ENABLE_SHIFT_PER_TARGET) +\
+	(4 * ((source) / 32)))
+#define PLIC_THRESHOLD(base, hart) \
+	((base) + PLIC_THRESHOLD_OFFSET +\
+	SHIFT_U32(hart, PLIC_THRESHOLD_SHIFT_PER_TARGET))
+#define PLIC_COMPLETE(base, hart) \
+	((base) + PLIC_CLAIM_OFFSET +\
+	SHIFT_U32(hartid, PLIC_CLAIM_SHIFT_PER_TARGET))
+#define PLIC_CLAIM(base, hart) \
+	PLIC_COMPLETE(base, hart)
+
+register_phys_mem_pgdir(MEM_AREA_IO_SEC, PLIC_BASE, PLIC_REG_SIZE);
+
+static inline bool plic_is_pending(struct plic_data *pd, uint32_t source)
+{
+	uint32_t current  = io_read32(PLIC_PENDING(pd->plic_base, source));
+
+	return ((current & BIT(source % 32)) != 0);
+}
+
+static inline void plic_set_pending(struct plic_data *pd, uint32_t source)
+{
+	uint32_t current  = io_read32(PLIC_PENDING(pd->plic_base, source));
+
+	current |= BIT(source % 32);
+	io_write32(PLIC_PENDING(pd->plic_base, source), current);
+}
+
+static inline void plic_enable_interrupt(struct plic_data *pd, uint32_t source)
+{
+	uint32_t hartid = get_core_pos();
+	uint32_t current;
+
+	current = io_read32(PLIC_ENABLE(pd->plic_base, source, hartid));
+	current |= BIT(source & 0x1f);
+	io_write32(PLIC_ENABLE(pd->plic_base, source, hartid), current);
+}
+
+static inline uint32_t plic_get_interrupt_enable(struct plic_data *pd,
+						 uint32_t source)
+{
+	uint32_t hartid = get_core_pos();
+	uint32_t current;
+
+	current = io_read32(PLIC_ENABLE(pd->plic_base, source, hartid));
+	current = current >> (source & 0x1f);
+	return current;
```

Comment:
Since this API tests only one bit, it should be:
```
current = io_read32(PLIC_ENABLE(pd->plic_base, source, hartid));
return 1 & (current >> (source & 0x1f));
```